### PR TITLE
Task 27 forgot password

### PR DESCRIPTION
**Pull request by OlegPanasyuk**

### CODE REVIEW

File: api/services/auth.js [hunk 1]
```diff
@@ -136,28 +141,30 @@ async function registration({ firstName, lastName, email, password, token }) {
...
-    return info;
+    let user = await db.Users.findOne({ where: { email } });
+    if (user) {
+      let linkId = (await hashHelpers.createHash(email));
+      let forgotPassword =
+        (await db.ForgotPassword.findOrCreate({
+          where: { linkId },
+          defaults: { UserId: user.id },
+        }));
+      let linkId2 = encodeURIComponent(linkId);
```
Comment (slavasgod) [inline api/services/auth.js]: Use sensible variables, what is `linkId2`?

File: api/services/auth.js [hunk 2]
```diff
@@ -136,28 +141,30 @@ async function registration({ firstName, lastName, email, password, token }) {
...
-      v && linkId && forgotPassword
-        ? await emailHelpers.sendEmail(
-            email,
-            `Follow link http://localhost:3000/reset?id=${linkId2}`,
-            `http://localhost:3000/reset?id=${linkId2}`
-          )
-        : { message: 'User not founded', status: 400 };
-    console.log(nodemailer.getTestMessageUrl(info));
-    return info;
+    let user = await db.Users.findOne({ where: { email } });
```
Comment (slavasgod) [inline api/services/auth.js]: In such cases better to use const

File: api/services/auth.js [hunk 3]
```diff
@@ -136,28 +141,30 @@ async function registration({ firstName, lastName, email, password, token }) {
...
+              email,
+              `Follow link ${HOST}/reset?id=${linkId2}`,
+              `${HOST}/reset?id=${linkId2}`
+            ));
+      console.log(nodemailer.getTestMessageUrl(info), ' ');
+      return info;
+    }
+    if(!user) {
+      throw new errors.UserNotFoundError();
+    }
```
Comment (slavasgod) [inline api/services/auth.js]: This if condition is not needed. You have return in your previous condition, so here you can just throw an error.
Maybe it's better to migrate this condition after your first request to db, and remove previous condition

File: api/services/auth.js [hunk 4]
```diff
@@ -166,27 +173,37 @@ const resetPasswordApprove = async ({ linkId }) => {
...
 };
 
+/**
+ * TODO: It needs to use Transactions
+ * @param {*} param0 
+ */
+
 const resetPassword = async ({ password, linkId }) => {
   try {
     const linkId2 = decodeURIComponent(linkId);
```
Comment (slavasgod) [inline api/services/auth.js]: Use sensible variables...

File: api/services/auth.js (hunk 1 above)
Comment (OlegPanasyuk) [reply api/services/auth.js]: it ls devilry. I'll fix it

File: api/services/auth.js [hunk 5]
```diff
@@ -1,9 +1,14 @@
 const nodemailer = require('nodemailer');
 const { hashHelpers, jwtHelpers, emailHelpers } = require('../helpers');
 const { ErrorHandler } = require('../middlewares/errorHandler');
+const ResetPasswordApproveError = require('../services/errorHandlers/resetPasswordApproveError');
+const ResetPasswordError = require('../services/errorHandlers/resetPasswordError');
+
 const errors = require('./errorHandlers/index');
 const db = require('../../database');
 
+const HOST = process.env.CL_HOST || 'http://localhost:3000';
```
Comment (slavasgod) [inline api/services/auth.js]: Why you make this variable global? I think it is not necessary to make this condition.

File: api/services/auth.js [hunk 6]
```diff
@@ -1,9 +1,14 @@
 const nodemailer = require('nodemailer');
 const { hashHelpers, jwtHelpers, emailHelpers } = require('../helpers');
 const { ErrorHandler } = require('../middlewares/errorHandler');
+const ResetPasswordApproveError = require('../services/errorHandlers/resetPasswordApproveError');
+const ResetPasswordError = require('../services/errorHandlers/resetPasswordError');
+
```
Comment (slavasgod) [inline api/services/auth.js:5]: Why your import this 2 errors from different files? We have index.js in our error folder, so you can import from there

File: api/controllers/auth.js [hunk 7]
```diff
@@ -77,18 +81,27 @@ const resetApprovementPassword = async (req, res, next) => {
       });
     }
   } catch (e) {
-    next(e);
+    throw new error.ResetPasswordApproveError(e.message);
   }
 };
 
 const resetPassword = async (req, res, next) => {
```
Comment (slavasgod) [inline api/controllers/auth.js]: Remove next from parameters

File: api/controllers/auth.js [hunk 8]
```diff
@@ -77,18 +81,27 @@ const resetApprovementPassword = async (req, res, next) => {
...
-    const v = await passwordSchema.validate({ password, passwordConfirm });
-    const a = v && (await services.resetPassword({ password, linkId }));
-    res.status(200).send(a);
-  } catch (e) {
-    next(e);
+  if (!linkId || !password || !passwordConfirm) {
+    throw new errors.ResetPasswordError('Empty params');
+  }
+  if (linkId && password && passwordConfirm) {
+    try {
```
Comment (slavasgod) [inline api/controllers/auth.js]: It is not needed to use both of this condition.

**Review by slavasgod — CHANGES_REQUESTED** [review]

File: api/services/auth.js (hunk 2 above)
Comment (OlegPanasyuk) [reply api/services/auth.js]: Ok. Done

File: api/services/auth.js (hunk 4 above)
Comment (OlegPanasyuk) [reply api/services/auth.js]: done

File: api/services/auth.js (hunk 5 above)
Comment (OlegPanasyuk) [reply api/services/auth.js]: I don't want to left the address of the server in code. It's not global it's a variable of environment

File: api/services/auth.js (hunk 6 above)
Comment (OlegPanasyuk) [reply api/services/auth.js:5]: Ok. Done

File: api/services/auth.js (hunk 3 above)
Comment (OlegPanasyuk) [reply api/services/auth.js]: Done

File: api/controllers/auth.js (hunk 8 above)
Comment (OlegPanasyuk) [reply api/controllers/auth.js]: Done

File: api/controllers/auth.js (hunk 7 above)
Comment (OlegPanasyuk) [reply api/controllers/auth.js]: Done

File: api/routes/auth.js [hunk 9]
```diff
@@ -23,5 +23,12 @@ router.get(
   controllers.socialLogin
 );
 
+router.post('/reset', controllers.reset);
+
+router.post('/reset/:linkId', controllers.resetApprovementPassword);
+
+router.post('/resetPassword', controllers.resetPassword);
+
```
Comment (slavasgod) [inline api/routes/auth.js:31]: We don't name routes in camelCase

File: api/services/auth.js [hunk 10]
```diff
@@ -61,6 +70,81 @@ const registration = async ({ firstName, lastName, email, password }) => {
...
+        user && {
+          status: 200,
+          message: 'Password updated',
+        }
+      );
+    }
+    if (!link) {
+      throw new ResetPasswordError();
+    }
+  } catch (e) {
```
Comment (slavasgod) [inline api/services/auth.js:144]: Remove `if`

File: api/services/errorHandlers/resetPasswordError.js [hunk 11]
```diff
@@ -0,0 +1,9 @@
+const ApplicationError = require('./applicationError');
+
+class ResetPasswordError extends ApplicationError {
+    constructor(message){
+        super(message || 'Link not found', 400);
+    }
+}
```
Comment (slavasgod) [inline api/services/errorHandlers/resetPasswordError.js:7]: Make common error for reset password

File: api/services/auth.js [hunk 12]
```diff
@@ -61,6 +70,81 @@ const registration = async ({ firstName, lastName, email, password }) => {
...
+          `Follow link ${HOST}/reset?id=${linkId}`,
+          `${HOST}/reset?id=${linkId}`
+        ));
+      console.log(nodemailer.getTestMessageUrl(info), ' ');
+      return info;
+    }
+    throw new errors.UserNotFoundError();
+  } catch (e) {
+    throw new Error(e.message);
+  }
```
Comment (slavasgod) [inline api/services/auth.js:105]: Throw custom error

File: api/services/auth.js [hunk 13]
```diff
@@ -61,6 +70,81 @@ const registration = async ({ firstName, lastName, email, password }) => {
...
+      let linkId = await hashHelpers.createHash(email);
+      const forgotPassword = await db.ForgotPassword.findOrCreate({
+        where: { linkId },
+        defaults: { UserId: user.id },
+      });
+      linkId = encodeURIComponent(linkId);
+      const info =
+        linkId &&
+        forgotPassword &&
+        (await emailHelpers.sendEmail(
```
Comment (slavasgod) [inline api/services/auth.js]: Remove this check

**Review by slavasgod — CHANGES_REQUESTED** [review]

File: api/routes/auth.js (hunk 9 above)
Comment (OlegPanasyuk) [reply api/routes/auth.js:31]: Done

File: api/services/auth.js (hunk 10 above)
Comment (OlegPanasyuk) [reply api/services/auth.js:144]: Done

File: api/services/auth.js (hunk 13 above)
Comment (OlegPanasyuk) [reply api/services/auth.js]: Done

File: api/services/errorHandlers/resetPasswordError.js (hunk 11 above)
Comment (OlegPanasyuk) [reply api/services/errorHandlers/resetPasswordError.js:7]: Done

File: api/services/auth.js (hunk 12 above)
Comment (OlegPanasyuk) [reply api/services/auth.js:105]: Done

File: api/controllers/auth.js [hunk 14]
```diff
@@ -17,13 +24,70 @@ const socialLogin = async (req, res) => {
...
+    });
+    const answer =
+      validation && (await services.resetPassword({ password, linkId }));
+    res.status(200).send(answer);
+  } catch (e) {
+    throw new errors.ResetPasswordError(e.message);
+  }
+};
+
+const googleLogin = async (req, res) => {
```
Comment (melonholy) [inline api/controllers/auth.js]: pull new dev branch and remove this function before merge

File: api/helpers/emailHelper.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+  // send mail with defined transport object
+  return transporter.sendMail({
+    from: '"Fred Foo 👻" <foo@example.com>', // sender address
+    to: address, // list of receivers
+    subject: 'Hello ✔', // Subject line
+    text, // plain text body
+    html: '<a href="' + link + '">ЛИНКА</a>', // html body
+  });
+  //   console.log('Message sent: %s', info.messageId);
```
Comment (melonholy) [inline api/helpers/emailHelper.js]: remove comments from git

File: api/services/auth.js [hunk 16]
```diff
@@ -1,7 +1,16 @@
...
+const { ErrorHandler } = require('../middlewares/errorHandler');
+const {
+  ResetPasswordApproveError,
+  ResetPasswordError,
+} = require('./errorHandlers/index');
+
 const errors = require('./errorHandlers/index');
 const db = require('../../database');
 
+const HOST = process.env.CL_HOST || 'http://localhost:3000';
```
Comment (melonholy) [inline api/services/auth.js]: we are using another env variable, pls check skype to get the newest .env file

File: database/v1/index.js [hunk 17]
```diff
@@ -32,4 +32,8 @@ Object.keys(db).forEach(modelName => {
 db.sequelize = sequelize;
 db.Sequelize = Sequelize;
 
+db.sequelize.sync({ force: false }).then(data => {
```
Comment (melonholy) [inline database/v1/index.js:35]: Are we still using v1 database? If we don't pls remove it

File: package.json [hunk 18]
```diff
@@ -19,6 +19,8 @@
     "express": "^4.17.1",
     "express-async-errors": "^3.1.1",
     "jsonwebtoken": "^8.5.1",
+    "jwt-decode": "^2.2.0",
```
Comment (melonholy) [inline package.json]: we removed this library from the project, please pull new dev branch

**Review by melonholy — CHANGES_REQUESTED** [review]

File: api/controllers/auth.js (hunk 14 above)
Comment (OlegPanasyuk) [reply api/controllers/auth.js]: Done

File: api/helpers/emailHelper.js (hunk 15 above)
Comment (OlegPanasyuk) [reply api/helpers/emailHelper.js]: Done

File: api/services/auth.js (hunk 16 above)
Comment (OlegPanasyuk) [reply api/services/auth.js]: Done

File: database/v1/index.js (hunk 17 above)
Comment (OlegPanasyuk) [reply database/v1/index.js:35]: Let's remove this DB in the next iteration.

File: package.json (hunk 18 above)
Comment (OlegPanasyuk) [reply package.json]: Done

**Review by slavasgod — APPROVED** [review]

**Review by melonholy — APPROVED** [review]

**Pull request merged**